apollo-federation/src/connectors/json_selection/apply_to.rs

```diff
@@ -723,83 +776,106 @@ impl ApplyToInternal for WithRange<PathList> {
-                .map_or_else(
-                    || {
+            PathList::Method(method_name, method_args, tail) => {
+                if let Some(method) = ArrowMethod::lookup(method_name) {
+                    // Before connect/v0.3, we did not consult method.shape at
+                    // all, and instead returned Unknown. Since this behavior
+                    // has consequences for URI validation, the older behavior
+                    // is preserved/retrievable given ConnectSpec::V0_2/earlier.
+                    if context.spec() < ConnectSpec::V0_3 {
+                        (
+                            Shape::unknown(method_name.shape_location(context.source_id())),
+                            None,
+                        )
+                    } else {
+                        (
+                            method.shape(
+                                context,
+                                method_name,
+                                method_args.as_ref(),
+                                input_shape,
+                                dollar_shape.clone(),
+                            ),
+                            Some(tail),
+                        )
+                    }
+                } else {
+                    (
                         Shape::error(
                             format!("Method ->{} not found", method_name.as_str()),
                             method_name.shape_location(context.source_id()),
-                        )
-                    },
-                    |_method| Shape::unknown(method_name.shape_location(context.source_id())),
-                ),
+                        ),
+                        None,
+                    )
+                }
+            }
 
             PathList::Selection(selection) => {
-                selection.compute_output_shape(context, input_shape, dollar_shape)
+                if input_shape.is_none() {
+                    return Shape::error(
+                        format!(
+                            "Selection cannot be applied to {}",
+                            input_shape.pretty_print()
+                        ),
+                        self.shape_location(context.source_id()),
+                    );
+                }
+
+                (
+                    selection.compute_output_shape(context, input_shape, dollar_shape.clone()),
+                    None,
+                )
             }
 
-            PathList::Empty => input_shape,
+            PathList::Empty => (input_shape, None),
+        };
+
+        if let (false, Some(tail)) = (current_shape.is_none(), tail) {
+            tail.compute_output_shape(context, current_shape, dollar_shape)
+        } else {
+            // If there is no tail or current_shape was None, path evaluation
+            // returns current_shape immediately.
+            current_shape
         }
```

Comment:
The new short-circuit check at the top means that we don't have to double-check `none` here, right? So a simpler check would be equivalent?


```suggestion
        if let Some(tail) = tail {
            tail.compute_output_shape(context, current_shape, dollar_shape)
        } else {
            current_shape
        }
```

Reply:
Yeah! Let's lean into that base case.